### PR TITLE
Add CoRe optimizer

### PR DESCRIPTION
**Pull request by eckhoffm**
Implements the Continual Resilient (CoRe) optimizer which shows impressive performance in a broad range of machine learning applications.
For further details regarding the algorithm and its performance we refer to the papers "Lifelong Machine Learning Potentials" (https://doi.org/10.1021/acs.jctc.3c00279) and "CoRe Optimizer: An All-in-One Solution for Machine Learning" (https://arxiv.org/abs/2307.15663).

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by jbschlosser** [comment]
@janeyx99 leaving this one to you :)

File: torch/optim/core.py [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+        params_decay = [-weight_decay[i % n_weight_decay] * param_updates[i].abs() * grouped_params[i]
+                        for i in range(len(grouped_step_sizes))]
+        torch._foreach_add_(grouped_params, params_decay)
+
+        # update parameters
+        torch._foreach_add_(grouped_params, param_updates, alpha=-1.0)
+
+        # update previous gradients
+        for i in range(len(grouped_prevs_1)):
+            grouped_prevs_1[i].copy_(grouped_grads[i])
```
Comment (mlazos) [inline torch/optim/core.py:561]: we have a `foreach_copy_` op that can speed this up.

**Comment by eckhoffm** [comment]
> Wow thanks for bringing this up and for taking the time to write the implementation (even with foreach :0). However, it is typically safer to create an issue proposing a new algorithm before jumping into implement it, as PyTorch Core has some pretty strict rules when it comes to adding a new Optimizer or Module: https://github.com/pytorch/pytorch/wiki/Developer-FAQ#i-have-a-new-function-or-feature-id-like-to-add-to-pytorch-should-it-be-in-pytorch-core-or-a-library-like-torchvision
> 
> In particular, this work might be a bit too new to be considered to be added into core just yet, and the way to go may be to put it in a separate library for now. Don't hesitate to let us know if we could provide any other low level features to help CoRe be even more performant.

Thank you very much for the fast response. I can understand this policy, even though I believe that it would be a great opportunity for many users to try the CoRe optimizer also for their projects.

**Comment by janeyx99** [comment]
@eckhoffm While visibility will be different, hosting it as a part of a third library should allow for users to start trying it. Then, once it’s battle tested with others’ proven results, we can reconsider adding CoRe to Core :)

**Pull request closed without merging**